rust/src/editor.rs

```diff
@@ -177,6 +178,28 @@ impl Editor {
         }
     }
 
+    fn cursor_start(&mut self) {
+        let start = self.view.sel_min() - self.col;
+        self.set_cursor(start, true);
+    }
+
+    fn cursor_end(&mut self) {
+        let current = self.view.sel_max();
+        let rope = self.text.clone();
+        let mut cursor = Cursor::new(&rope, current);
+        let mut new = current;
+        // Advance this cursor as this is considered a boundary.
+        if new == 0 {
```

Comment:
The return value of `cursor.next()` is the offset in the buffer of the next line (paragraph) boundary. Note, that'll be after the `'\n'`.

I think the logic you want is:

* Do `cursor.next()` from current.
* If you're not on a boundary (`is_boundary`), you're at the end of the file and there's no newline at the end. Use that.
* If you are on a boundary, go back one grapheme (see end of `line_col_to_offset` in view.rs)

Does that help?